include/internal/hashtable.h

```diff
@@ -0,0 +1,319 @@
+                                      vtype **olddata);                        \
+vtype *ossl_ht_##name##_##vtype##_from_value(HT_VALUE *v);                     \
+vtype *ossl_unused ossl_ht_##name##_##vtype##_get(HT *h,                       \
+                                                  HT_KEY *key,                 \
+                                                  HT_VALUE **v);               \
+HT_VALUE *ossl_ht_##name##_##vtype##_to_value(vtype *data, HT_VALUE *v);       \
+int ossl_ht_##name##_##vtype##_type(HT_VALUE *h);                              \
+
+/*
+ * Helper function to construct case insensitive keys
+ */
+static inline void ossl_ht_strcase(char *tgt, const char *src, int len)
+{
+    int i;
+#if defined(CHARSET_EBCDIC) && !defined(CHARSET_EBCDIC_TEST)
+    const long int case_adjust = ~0x40;
+#else
+    const long int case_adjust = ~0x20;
+#endif
+
+    if (src == NULL)
+        return;
+
+    for (i = 0; src[i] != '\0' && i < len; i++)
+        tgt[i] = case_adjust & src[i];
+}
+
+/*
+ * Create a new hashtable
+ */
+HT *ossl_ht_new(HT_CONFIG *conf);
+
+/*
+ * Frees a hash table, potentially freeing all elements
+ */
+void ossl_ht_free(HT *htable);
+
+/*
+ * Empties a hash table, potentially freeing all elements
+ */
+int  ossl_ht_flush(HT *htable);
+
+/*
+ * Inserts an element to a hash table, optionally returning
+ * replaced data to caller
+ */
+int ossl_ht_insert(HT *htable, HT_KEY *key, HT_VALUE *data,
+                   HT_VALUE **olddata);
+
+/*
+ * Deletes a value from a hash table, based on key
+ * results provides information about delete status
+ */
+#define HT_DEL_FLAG_DELETED            0x1
+#define HT_DEL_FLAG_VALUE_NOT_RETURNED 0x2
+#define HT_DEL_FLAG_DEL_NOT_SUPPORTED  0x4
+HT_VALUE *ossl_ht_delete(HT *htable, HT_KEY *key, uint32_t *results);
+
+/*
+ * Returns number of elements in the hash table
+ */
+size_t ossl_ht_count(HT *htable);
+
+/*
+ * Iterates of the hash table calling cb for each element
+ */
+void ossl_ht_foreach(HT *htable, void (*cb)(HT_VALUE *obj, void *arg),
+                     void *arg);
+
+/*
+ * Variant of foreach, but aborts the loop when cb returns 0
+ */
+void ossl_ht_foreach_until(HT *htable, int (*cb)(HT_VALUE *obj, void *arg),
+                           void *arg);
+/*
+ * Returns a list of elements in a hash table based on
+ * filter function return value
+ */
+HT_VALUE_LIST *ossl_ht_filter(HT *htable, size_t max_len,
+                              int (*filter)(HT_VALUE *obj, void *arg),
+                              void *arg);
+/*
+ * Frees the list returned from ossl_ht_filter
+ */
+void ossl_ht_value_list_free(HT_VALUE_LIST *list);
+
+/*
+ * Fetches a value from the hash table, based
+ * on key
+ */
+HT_VALUE *ossl_ht_get(HT *htable, HT_KEY *key);
+
+/*
+ * Drops the reference count on hash table values
+ * returned from ossl_ht_filter, ossl_ht_get, and
+ * ossl_ht_insert, potentially freeing the value
+ */
+void ossl_ht_put(HT_VALUE *value);
```

Comment:
This is a hashtable. Why is it conflating and integrating reference counters? I think this is a very bad idea. Also no getter which isn't mutating of this refcount.

Reply:
because data can be shared.  My intent was to remove the need for external locking of a table.  When you insert a value, the data that is inserted becomes owned by the table.  Lookups can run concurrent to deletions, and to do so safely requires us to know on return from the lookup that the data retrieved will remain stable.  the only way to do that is to either (a) lock the table externally or (b) refcount the objects retrieved.

Note: There is a no-refcounting mode of this hash table documented in which the caller still maintains the responsibility to ensure protection of concurrent operations.

Reply:
Note, in the most recent push, I've exposed a locking api for the table that allows for user directed locking (or not if lockless behavior is feasible).  This lets us get rid of several config options, reduces code size for the hash table, and lets us eliminate reference counting inside the table entirely